src/os/bluestore/BlueFS.cc

```diff
@@ -1239,17 +1270,23 @@ void BlueFS::_fsync(FileWriter *h)
 {
   dout(10) << __func__ << " " << h << " " << h->file->fnode << dendl;
   _flush(h, true);
-  _flush_wait(h);
-  if (h->file->dirty) {
-    dout(20) << __func__ << " file metadata is dirty, flushing log on "
-	     << h->file->fnode << dendl;
-    _flush_log();
-    assert(!h->file->dirty);
+  uint64_t old_dirty_seq = h->file->dirty_seq;
+  lock.unlock();
+  wait_for_aio(h);
```

Comment:
There is a problem here, Thread T1 will finish _flush(updates log file) and waits for the IO to complete. Thread T2 can grab the lock, and do _flush(update the log file), releases the lock. T1 will grab and flushes the log file contains T1 contents + T2 contents. If T2 IO fails now, we will have log containing T2 updates.